x-pack/plugins/task_manager/server/queries/oldest_idle_action_task.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { ElasticsearchClient } from 'kibana/server';
+
+/**
+ * Returns the timestamp of the oldest action task that may still be executed.
+ * Useful for cleaning up related objects that may no longer be needed.
+ */
+export const getOldestIdleActionTask = async (
+  client: Pick<ElasticsearchClient, 'search'>,
+  taskManagerIndex: string
+): Promise<number | string> => {
+  let oldestAlertAction: number | string = Date.now();
+  const results = await client.search<{ task: { runAt: string } }>({
+    size: 1,
+    index: taskManagerIndex,
+    body: {
+      sort: [{ 'task.runAt': { order: 'asc' } }],
+      query: {
+        bool: {
+          filter: {
+            bool: {
+              must: [
+                {
+                  terms: {
+                    'task.taskType': [
+                      'actions:.email',
```

Comment:
We could move these to a constants file to make them easier to find in the long term but we can hand that off to plugin owners to decide.

Reply:
@mikecote is there a registry of taskTypes that we should use for this so it dynamically picks up any new actions?

Reply:
>@mikecote is there a registry of taskTypes that we should use for this so it dynamically picks up any new actions?

There is a registry that is part of the task manager's plugin. Any plugin can call the task manager's setup contract and register a task type. Afterwards, the registry contains all the taskTypes. I'm not sure if it would be accessible from the `deleteOnUpgrade` function by the time it is called?

Reply:
> Any plugin can call the task manager's setup contract and register a task type. Afterwards, the registry contains all the taskTypes. I'm not sure if it would be accessible from the `deleteOnUpgrade` function by the time it is called?

Yeah it should be since this hook isn't called until after `setup` completes. I'll add this as a potential follow up, I don't think it should block getting this fix in for the next BC.